pipelines/common/src/main/java/org/openmrs/analytics/JdbcConnectionPools.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2020-2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openmrs.analytics;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import com.mchange.v2.c3p0.ComboPooledDataSource;
+import java.beans.PropertyVetoException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.sql.DataSource;
+import org.openmrs.analytics.model.DatabaseConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JdbcConnectionPools {
+  private static final Logger log = LoggerFactory.getLogger(JdbcConnectionPools.class);
+
+  private static final ConcurrentMap<DataSourceConfig, DataSource> dataSources =
+      new ConcurrentHashMap<>();
+
+  // This class should not be instantiated!
+  private JdbcConnectionPools() {}
+
+  /**
+   * Creates a new connection pool for the given config or return an already created one if one
+   * exists. The pool size parameters are just hints and won't be used if a pool already exists for
+   * the given configuration. This method is to impose a Singleton pattern for each config.
+   *
+   * @param config the JDBC connection information
+   * @param initialPoolSize initial pool size if a new pool is created.
+   * @param jdbcMaxPoolSize maximum pool size if a new pool is created.
+   * @return
+   */
+  public static synchronized DataSource getPooledDataSource(
```

Comment:
I know our use case does not need this synchronization to be performant. But we can follow best practices as suggested here: https://medium.com/@kevalpatel2106/how-to-make-the-perfect-singleton-de6b951dfdb0

Reply:
Thanks for the link; can you point out specific parts that you think should be covered in this class but are not? I took a look at the recommendations in that link; what might be relevant (but not covered) is the reflection related part which I don't  think worth the added complexity (if the library users really want to shoot themselves in the foot with reflection, let them :-) ). The serialization related ones do not apply here. And the double-check/volatile part I have addressed in the next comment.

Reply:
I wanted to point towards double-check and serialization. How is that serialization related things are not applied here? Can we not end up creating different instances? Lets have unit test cases to make sure that we have singletons in place.

Reply:
Serialization of which object are you concerned about, `dataSources` or the individual `DataSource` objects in it? We do not serialize the `DataSource` objects and I want to keep it that way. I added a note to the javadoc above to clarify this as I think it is a key design point of this class. If you are concerned about `dataSources` map, it should not be serialized/deserialized either (for the same `DataSource` reason). I changed this class to a real singleton which does _not_ implement `Serializable` to make this very clear.

Reply:
Thanks for the clarification!